### PR TITLE
[JENKINS-38778] Fix leaked node without releasing

### PR DESCRIPTION
**Pull request by pjanouse**

### CODE REVIEW

**Comment by pjanouse** [comment]
Note: I know the reporting of "ProvisioningFailedException: No Foreman resources available" isn't optimal, however we don't have another option currently. It is bad practice to do a provisioning synchronously, we really need to do it async and in that case we need to have a 'signal' how to tell "well, had had the plan to provision a X, but it failed for whatever reason", it should be 'AbortException' in the future.

I've inspired in openstack-cloud-plugin, but it will need to be rewritten in this part too. Unfortunately the change has to be started from `hudson.slaves.NodeProvisioner` and it takes a bit more time until merged to upstream core.

**Comment by pjanouse** [comment]
FYI: For Jenkins upcoming change - see [JENKINS-38903] (https://issues.jenkins-ci.org/browse/JENKINS-38903), I'll working on it soon.

**Comment by pjanouse** [comment]
Rebased to resolve merge conflict.

**Comment by scoheb** [comment]
Why is throwing a ProvisioningFailedException better?

**Comment by pjanouse** [comment]
It isn't better, it is there only for the fact that OpenStack plugin threats similar situation currently in the same way. I expect almost all such the situations should end up in `AbortException` later.

**Comment by scoheb** [comment]
Has this change demonstrated that the problem is fixed?

**Comment by pjanouse** [comment]
We can't return `null` as the result of the `Future`  because NPE occurred later [here] (https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/slaves/NodeProvisioner.java#L225), resp. inside addNode().

**Comment by pjanouse** [comment]
Any Exception like AbortException/IOException is confusing currently because Foreman was node never ever created in such a situation, that's the reason for JENKINS-38903.

**Comment by pjanouse** [comment]
This PR at least fixes:
- async node provision process
- fixed occurred NPE in the core (not returning a NULL value)
- properly signal that we can provision, but we run out of resources though

Proposed change in the core is compatible with this, however we should use `AbortException` later (when it will be properly handled in the `NodeProvisioner`

**Comment by scoheb** [comment]
I fail to see how this fixes "async node provision process"

**Comment by pjanouse** [comment]
Your original code waited for a result from `Future` *before* adding to `PlannedNode` structure and return it. Bear in mind Jenkins goes through all node providers (one thread) and tries to acquire all necessary resources - you blocked that procedure and thread until right provisioning Foreman node is ready to use.

My fix passes the operation over `PlannedNode` collection to `NodeProvisioner` which goes through all PlannedNode and processes only already finished nodes (`future.isDone()`). The OpenStack provider plugin does the same.

**Comment by pjanouse** [comment]
I realized we can signal via `AbortException` already, so changed the code to not introduce another type of Exception.

**Comment by pjanouse** [comment]
Hmmm, just realized I'm still in the trouble here because Disposable object disappeared over night without releasing my nodes on Foreman side. 

Please hold on to process this PR until I find the cause.

File: src/main/java/com/redhat/foreman/ForemanComputer.java [hunk 1]
```diff
@@ -63,7 +65,12 @@ static void terminateForemanComputer(Computer c) throws IOException, Interrupted
             Node node = fc.getNode();
             if (node != null) {
                 ForemanSharedNode sharedNode = (ForemanSharedNode)node;
+                if (sharedNode.getChannel() != null) {
+                    sharedNode.getChannel().close();
+                }
                 sharedNode.terminate();
+                Jenkins.getInstance().removeNode(sharedNode);
```
Comment (scoheb) [inline src/main/java/com/redhat/foreman/ForemanComputer.java]: `AbstractCloudSlave.terminate()` calls `Jenkins.getInstance().removeNode(this);` already.

**Pull request closed without merging**